docs/adr/0016-use-graphql-and-apollo-for-settings-redesign.md

```diff
@@ -0,0 +1,135 @@
+- Since only required data will be requested, the data returned won't be overfetched like it is in traditional REST architecture, resulting in a smaller payload.
+- GQL offers a flexible and delcarative approach to data - both FE and BE developers can focus on describing the data rather than implementing and optimizing numerous REST endpoints. This makes it painfully clear what's being sent and received between client and server which can improve visibility into what's happening in components.
+- GraphQL and Apollo have gained a lot of popularity over the years and now have substantial company and community support.
+- This stack promotes strong end-to-end typing.
+- Apollo offers great features out of the box like caching, error handling, optimistic rendering, and nice integration tools for React (and even provides React hooks).
+- The React Settings application will only have to manage interacting with a single endpoint and Apollo Client makes managing the data in our React app straightforward. Apollo Client creates its own internal Redux store under the hood and allows us to store local data in the Apollo cache, the same place that data from network requests will be stored.
+- Apollo also provides a "GraphQL playground" which is a graphical, interactive, in-browser GQL IDE. It not only provides easy to access schemas (and is "self-documenting" in a way), but it allows developers to write queries or mutations to send to the server so the developer can verify it behaves as expected. This can then be copied directly over to where it needs to be used in the application.
+- CRUD operations have been simplified to simply queries and mutations. CRUD in GraphQL becomes "MQMM" and developers don't have to think about a C/U/D differentiation outside of writing the mutation for them.
+- GraphQL makes it easier to evolve APIs over time. It provides a way to deprecate schema members and discourage usage of certain parts of APIs with custom messaging.
+- GraphQL is database agnostic.
+- GraphQL queries are easy to write and understand.
+
+##### Cons
+
+- The learning curve that comes with introducing new tech.
+- Initial setup time of another service (`graphql-api-server`).
+- While GQL caching can be done at the database or client level though Apollo, GQL doesn't rely on HTTP caching methods.
+- While GQL has been around long enough to gain excellent community support, REST has been around with established patterns for even longer.
+- Queries always return a 200 status code, even if the query errored. If the query is unsuccessful, the response will instead have an `errors` key with associated error messages and stacktrace. However, the error messaging can be quite detailed and include all the resolvers and refer to the exact part of the query at fault.
```

Comment:
Hmmm, I might even consider this a neutral. It's not that a consistent 200 status is bad, per se, it's just _different_ than what's expected from a traditional REST setup. After all, the network request is successful, it's just the database operation that might not be, right?

Relatedly, what happens if the GQL server can't be reached at all?

Reply:
Ah I missed this comment from this morning.

I actually had this in "Neutral" at first because of the level of detail I've read you can add to errors, but moved it into "Cons" because I feel yeah, it's not what you'd expect to happen. I'll move it back to neutral.

I'm reading, "Non-200 response codes indicate that some issue occurred at the HTTP transport layer, not the GraphQL layer." If it can't be reached, we'll just see a `404 not found`, I just killed the `fxa-admin-server` while trying to query in `fxa-admin-panel` to double check. I do wonder if we hit some `500` error in the auth-server and Apollo sends us back a `200` how easily we can find that error code or what exactly we'll see under `errors`, maybe @bbangert knows?

Reply:
This explanation makes sense to me. Thanks for checking!